### DISCUSSION TITLE
Custom Models

### DISCUSSION THREAD

**Discussion by DeFlanko** [xnorpx/blue-onyx#6] [Q&A] [answered]
Where do we find existing custom models? How do we make custom models?

i would like to make an attempt at how to even begin to train a custom model. For my use case is to have my front door cam and or overlooking cam identify Parcels (UPS, FedEx, USPS, Amazon)  -- all kinds of Boxes on doorstep. 

I understand ill need a plethora of images (at least 1k right?) of boxes and packages alike.  (that headache of curating images aside) -- is there an easy step by step guide on creating a ONNX file?

**Comment by xnorpx** [comment, marked as answer]
Here is how I would do it if I had the time and need for custom models.

https://github.com/lyuwenyu/RT-DETR/tree/main/rtdetrv2_pytorch

``` bash
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --master_port=9909 --nproc_per_node=4 tools/train.py -c path/to/config --use-amp --seed=0 &> log.txt 2>&1 &
```

Once you are happy with your training you convert it to ONNX: 

``` bash
python tools/export_onnx.py -c path/to/config -r path/to/checkpoint --check
```

Create a yaml file with the classes similar to: https://github.com/xnorpx/blue-onyx/blob/main/assets/coco_classes.yaml

Provide the custom model and the custom_classes.yaml to the cli.

Note currently the input to the model is hardcoded to 640x640. For other resolution we need to make this configurable. But I will wait doing that change until someone comes and says that they have a model requiring 320x320 or 1280x1280 model.
